### PR TITLE
Apache airflow dag runs 2

### PR DESCRIPTION
**Pull request by stichiboi**
## Hey, I just made a Pull Request!

Added a new column in the table to quickly view the latest DAG runs, plus a link to it if you want to have a deeper look.
image

Also:
1. Table columns are togglable
2. Fixed bug with turning on/off the DAGs

Check https://github.com/backstage/backstage/issues/12812 for discussions

Note: this is a duplicate of [this PR](https://github.com/backstage/backstage/pull/13644), which had some issues with the automated actions.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/apache-airflow/src/components/DagTableComponent/DagTableComponent.tsx [hunk 1]
```diff
@@ -136,8 +186,26 @@ type DagTableComponentProps = {
 
 export const DagTableComponent = (props: DagTableComponentProps) => {
   const { dagIds } = props;
+  const [dagsData, setDagsData] = useState<DagTableRow[] | []>([]);
```
Comment (freben) [inline plugins/apache-airflow/src/components/DagTableComponent/DagTableComponent.tsx:189]:
```suggestion
  const [dagsData, setDagsData] = useState<DagTableRow[]>([]);
```

you don't need that part

File: plugins/apache-airflow/src/components/DagTableComponent/DagTableComponent.tsx [hunk 2]
```diff
@@ -146,17 +214,25 @@ export const DagTableComponent = (props: DagTableComponentProps) => {
     return await apiClient.listDags();
   }, []);
 
+  const data = value?.map(el => ({
+    ...el,
+    id: el.dag_id, // table records require `id` attribute
+    dagUrl: `${apiClient.baseUrl}dag_details?dag_id=${el.dag_id}`, // construct path to DAG using `baseUrl`
+  }));
```
Comment (freben) [inline plugins/apache-airflow/src/components/DagTableComponent/DagTableComponent.tsx:221]: Shouldn't you be doing this work inside the useAsync above instead? Then the returned value would already be exactly what you need. And then the useEffect below could use just `[data]` as its dependency and not need the react-hooks/exhaustive-deps exclusion.

But then again - now, dagsData and setDagsData feels like they would be entirely unnecessary now, right? I mean, if what comes out of the useAsync is exactly what you need, there's no need to double-store it again in a separate useState holder. Could just use `value` directly instead.

File: plugins/apache-airflow/src/components/LatestDagRunsStatus/LatestDagRunsStatus.tsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+export const LatestDagRunsStatus = ({
+  dagId,
+  limit = 5,
+}: LatestDagRunsStatusProps) => {
+  const classes = useStyles();
+  const apiClient = useApi(apacheAirflowApiRef);
+  const { value, loading, error } = useAsync(async (): Promise<DagRun[]> => {
+    return await apiClient.getDagRuns(dagId, { limit });
+  }, []);
```
Comment (freben) [inline plugins/apache-airflow/src/components/LatestDagRunsStatus/LatestDagRunsStatus.tsx:109]: could shorten this to `useAsync(() => apiClient.getDagRuns(dagId, { limit }), [dagId, limit])` (note that I added a dependency on `dagId` and `limit` - technically, the request should be re-run if they change

File: plugins/apache-airflow/src/components/LatestDagRunsStatus/LatestDagRunsStatus.tsx [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+  if (loading) {
+    return (
+      <Box>
+        <CircularProgress />
```
Comment (freben) [inline plugins/apache-airflow/src/components/LatestDagRunsStatus/LatestDagRunsStatus.tsx:114]: Hm normally we only use `<Progress />` from `@backstage/core-components` as a matter of consistency in the interface

File: plugins/apache-airflow/src/components/DagTableComponent/DagTableComponent.tsx [hunk 5]
```diff
@@ -31,101 +31,151 @@ import Tooltip from '@material-ui/core/Tooltip';
...
+      onChangeColumnHidden={(column, hidden) => {
+        if (column.field) {
+          let newHiddenColumns: string[];
+          if (hidden) {
+            newHiddenColumns = hiddenColumns.concat(column.field);
+          } else {
+            newHiddenColumns = hiddenColumns.filter(v => v !== column.field);
+          }
+          setHiddenColumns(newHiddenColumns);
+          window.localStorage.setItem(
```
Comment (freben) [inline plugins/apache-airflow/src/components/DagTableComponent/DagTableComponent.tsx:173]: You could see usages of `storageApiRef` instead of using local storage here in the main repo - if you want this to persist for the current user across sessions and devices

File: plugins/apache-airflow/src/api/ApacheAirflowClient.ts [hunk 6]
```diff
@@ -97,11 +98,55 @@ export class ApacheAirflowClient implements ApacheAirflowApi {
...
+      }>(`/dags/${dagId}/dagRuns?${qs.stringify(searchParams)}`);
+      dagRuns.push(...response.dag_runs);
+
+      if (
+        dagRuns.length >= response.total_entries ||
+        dagRuns.length >= options.limit
+      ) {
+        break;
+      }
+      if (typeof searchParams.offset !== 'undefined') {
```
Comment (freben) [inline plugins/apache-airflow/src/api/ApacheAirflowClient.ts:143]: the string undefined, really? :) ok

File: plugins/apache-airflow/src/api/ApacheAirflowClient.ts [hunk 7]
```diff
@@ -97,11 +98,55 @@ export class ApacheAirflowClient implements ApacheAirflowApi {
...
+
+    for (;;) {
+      const response = await this.fetch<{
+        dag_runs: DagRun[];
+        total_entries: number;
+      }>(`/dags/${dagId}/dagRuns?${qs.stringify(searchParams)}`);
+      dagRuns.push(...response.dag_runs);
+
+      if (
+        dagRuns.length >= response.total_entries ||
```
Comment (freben) [inline plugins/apache-airflow/src/api/ApacheAirflowClient.ts:138]: feels like we need another break condition in here, as in when the source has no more data to give. Maybe when `response.dag_runs.length` or `response.total_entries` is zero or so?

File: .changeset/eighty-comics-suffer.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-apache-airflow': minor
```
Comment (freben) [inline .changeset/eighty-comics-suffer.md:2]: This could be a `patch` since no public interface changes and we're on 0.x

**Review by freben — COMMENTED** [review]
Nice, thanks! Some remarks herein

**Comment by stichiboi** [comment]
Pinging @loybin

File: plugins/apache-airflow/src/components/DagTableComponent/DagTableComponent.tsx (hunk 2 above)
Comment (LoyBin) [reply plugins/apache-airflow/src/components/DagTableComponent/DagTableComponent.tsx]: I put the map function into useAsync block, but the dagsData and setDagsData  are necessary because we need to update it when click the Paused switch

File: plugins/apache-airflow/src/components/LatestDagRunsStatus/LatestDagRunsStatus.tsx (hunk 4 above)
Comment (LoyBin) [reply plugins/apache-airflow/src/components/LatestDagRunsStatus/LatestDagRunsStatus.tsx:115]: <Progress /> is linearProgress, but in here we need circularProgress, and Backstage don't have a wrapper about it. According Backstage's convention, it's fine to use MUI components.

File: plugins/apache-airflow/src/api/ApacheAirflowClient.ts (hunk 6 above)
Comment (LoyBin) [reply plugins/apache-airflow/src/api/ApacheAirflowClient.ts]: `searchParams.offset` can't be undefined actually, we assign it a number value above.

**Comment by freben** [comment]
I think you got master merged into this one, or made some form of an odd rebase - there's a lot of conflicts and unwanted changes now.

**Comment by LoyBin** [comment]
> I think you got master merged into this one, or made some form of an odd rebase - there's a lot of conflicts and unwanted changes now.

@freben What can we do to solve it? It is still hanging in there.

**Comment by Rugvip** [comment]
Reopening to make sure this doesn't get lost. Some strategies for restoring:

Start by creating a back-up point, `git branch dag-runs-backup`, it's good to have this so you can go back to the current state at any point using `git reset --hard dag-runs-backup`.

Now on the current branch, you could try an interactive rebase, `git rebase -i master`. In the editor that pops up, delete the lines for any commits that you don't want, then save and exit. This could get you to the point you want.

Another option is if you can go back to a known good state, check the reflog, `git reflog apache-airflow-dag-runs-2`. You can jump to any commit using `git reset --hard <sha>`.

You can also create a new branch, and cherry-pick the commits over, starting at the root. In practice this is similar to the rebase approach.

If all else fails, create a new branch and copy over the changes from this one.

**Comment by LoyBin** [comment]
> Reopening to make sure this doesn't get lost. Some strategies for restoring:
> 
> Start by creating a back-up point, `git branch dag-runs-backup`, it's good to have this so you can go back to the current state at any point using `git reset --hard dag-runs-backup`.
> 
> Now on the current branch, you could try an interactive rebase, `git rebase -i master`. In the editor that pops up, delete the lines for any commits that you don't want, then save and exit. This could get you to the point you want.
> 
> Another option is if you can go back to a known good state, check the reflog, `git reflog apache-airflow-dag-runs-2`. You can jump to any commit using `git reset --hard <sha>`.
> 
> You can also create a new branch, and cherry-pick the commits over, starting at the root. In practice this is similar to the rebase approach.
> 
> If all else fails, create a new branch and copy over the changes from this one.

@Rugvip Thank you! I have tried rebase and reflog, but failed. It has too many commits between this branch and HEAD

I create a new branch, the new PR is https://github.com/backstage/backstage/pull/14506
This can be close now.
